---
title: "Analysis 10X dataset"
output:
  html_document:
    fig_caption: yes
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    number_sections: true
    theme: cerulean
editor_options:
  chunk_output_type: console
params:
  prefix: "combined"
  resDEG: "SCT_snn_res.0.25"
  percent.mito: 0.15
  cell.source: "human"
  harmonize: TRUE
  sample.names: !r c("KD_CF","KD_R")
---


```{r setup, include=FALSE, cache=FALSE}
library(future)
plan("multiprocess", workers = 8)
options(future.globals.maxSize = 8000 * 1024^2)

ifelse(dir.exists("/mnt/scc/project_workspace/"),
       cremwd <- "/mnt/scc/",
       cremwd <- "/restricted/projectnb/crem-bioinfo/")
master_path <- file.path(cremwd, "/project_workspace/18_09_25_kostas/calculations/analysis/")
calculations_path <- paste0(master_path, "combined_harmony/")
plots_path <- file.path(calculations_path, "plots/")
rds_path <- file.path(calculations_path, "rds/")
cache_path <- file.path(calculations_path, "cache/") # it must end in a slash    
dir.create(rds_path, recursive = T)
dir.create(plots_path, recursive = T)

knitr::opts_chunk$set(echo = FALSE, message = FALSE,error = FALSE,warning = FALSE, cache = TRUE, cache.path = cache_path, fig.width=9,fig.height=7, autodep=TRUE, 
collapse = FALSE, fig.path = plots_path, fig.keep = "all", comment = "##", dev = 
c("png", "pdf"), cache.lazy = FALSE)
options(tibble.print_max = 150)

library(tidyverse)
library(Seurat)
library(Matrix)
library(matrixStats) #rowVars
library(magrittr)
library(dplyr)
library(RColorBrewer)
library(kableExtra)
library(data.table) # getting counts
library(knitr)
library(clustree)
library(patchwork)
library(harmony)
library(MAST)

params <- list()
params$cell.source <- "human"
params$resDEG <- "SCT_snn_res.0.25"
params$harmonize = FALSE
```


```{r load_objects, eval=FALSE}
sc <- readRDS(paste0(rds_path, "sc.Rds"))
```

```{r load_and_merge_from_scratch}
seu.list <- list()

seu.list[[1]] <- readRDS("/restricted/projectnb/crem-bioinfo/project_workspace/19_10_29_kostas/calculations/analysis/KA_B2/rds/sc.Rds")  
seu.list[[1]]@project.name <- "KA_B2"
seu.list[[1]]$orig.ident <- "KA_B2"
DefaultAssay(seu.list[[1]]) <- "RNA" # make default RNA, so that SCT can be removed
seu.list[[1]] <- DietSeurat(seu.list[[1]], counts = TRUE, scale.data = FALSE, assays = "RNA")

seu.list[[2]] <- readRDS("/restricted/projectnb/crem-bioinfo/project_workspace/19_10_29_kostas/calculations/analysis/KA_C11/rds/sc.Rds")  
seu.list[[2]]@project.name <- "KA_C11"
seu.list[[2]]$orig.ident <- "KA_C11"
DefaultAssay(seu.list[[2]]) <- "RNA" # make default RNA, so that SCT can be removed
seu.list[[2]] <- DietSeurat(seu.list[[2]], counts = TRUE, scale.data = FALSE, assays = "RNA")

seu.list[[3]] <- readRDS("/restricted/projectnb/crem-bioinfo/project_workspace/18_09_25_kostas/calculations/analysis_B2.C11/sc.Rds")  
seu.list[[3]] <- UpdateSeuratObject(seu.list[[3]])
seu.list[[3]]@project.name <- "analysis_B2.C11"
seu.list[[3]]$orig.ident <- paste0("early_dev_d30_",seu.list[[3]]$orig.ident )
DefaultAssay(seu.list[[3]]) <- "RNA" # make default RNA, so that SCT can be removed
seu.list[[3]] <- DietSeurat(seu.list[[3]], counts = TRUE, scale.data = FALSE, assays = "RNA")

sc.pre <- merge(seu.list[[1]], y = seu.list[-1], project = "sc")
```

# QC filtering

```{r load_sc_qc, fig.height=4}
cat("UMI counts per cell")
summary(apply(GetAssayData(sc.pre, slot = "counts"), 2, sum))
cat("UMI counts per gene")
summary(apply(GetAssayData(sc.pre, slot = "counts"), 1, sum))

Idents(sc.pre) <- "orig.ident"

cat("Before filtering")
p <- VlnPlot(sc.pre, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0.1)
p <- p & ggmin::theme_powerpoint() & theme(axis.title.x = element_blank()) & NoLegend() & RotatedAxis()
p
ggsave(paste0(plots_path, "Vln.qc.pre.pdf"))
sc.pre
quantile.threshold <- (100 - ((dim(sc.pre)[2]/2)/1000)) / 100  # real cells are much less than detected barcodes. To estimate the duplet threshold more realistically, divide the barcode number by half

nGene.upper <- quantile(sc.pre$nFeature_RNA, probs = quantile.threshold)
nUMI.upper <- quantile(sc.pre$nCount_RNA, probs = quantile.threshold)
# nGene.upper <- 3500
# nUMI.upper <- 15000
cat("Doublet nGene threshold: ", nGene.upper)
cat("Doublet nUMI threshold: ", nUMI.upper)
keep.cells <- sc.pre[[]] %>% rownames_to_column() %>% dplyr::filter(nFeature_RNA < nGene.upper & nFeature_RNA > 800 & nCount_RNA < nUMI.upper & percent.mito < 25) %>% pull(rowname)
sc <- sc.pre[, keep.cells]
saveRDS(keep.cells, paste0(rds_path, "keep.cells.Rds"))

cat("After filtering doublets and degraded cells")
p <- VlnPlot(sc, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0.1)
p <- p & ggmin::theme_powerpoint() & theme(axis.title.x = element_blank()) & NoLegend() & RotatedAxis()
p
ggsave(paste0(plots_path, "Vln.qc.after.pdf"))


cat("After filtering doublets and degraded cells")
p <- VlnPlot(sc, features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), pt.size = 0.1)
p <- p & ggmin::theme_powerpoint() & theme(axis.title.x = element_blank()) & NoLegend() & RotatedAxis()
p
ggsave(paste0(plots_path, "Vln.qc.after.pdf"))
sc
```


```{r interesting_genes}
interesting_genes <- c(
  "NKX2-1", "EGFP_EXTRA", "TDTOMATO_EXTRA_IVS", "SFTA3", "SFTPB", "SOX2",        
  "AFP", "ALB",
  "TFF1", # gut/gastric
  "CDX2", # intestinal
  "SCGB3A2", # type 2 cells
  "TOP2A", "AURKB", "MKI67", "BIRC5", # cell cycle
  "TP63", "KRT5",
  "SCGB1A1",
  "SFTPC", "TFF2",
  "SOX9",
  "HP",
  "FOXJ1",
  "BMP3", "FOXP2", "CPM", "FGF9",
  "NKD1",
  "GIF", "CLDN18", "MUC6",
  "TF", "APOA1", "CYP4F3", "CYP3A7",
  "PAX9", "KRT14",
  "SFTA2", "AQP4",
  "AXIN2",
  "CFTR",
  "FOXI1", "ASLC3",
  "ABCA3")
firstup <- function(x) {
  x <- tolower(x)
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}
if(params$cell.source == "mouse") { interesting_genes <- firstup(interesting_genes) }
cat("Pre-selected genes: ", interesting_genes)
interesting_genes <- interesting_genes[interesting_genes %in% rownames(sc)]      
cat("Pre-selected genes expressed in current dataset: ", interesting_genes)      
```

# Dimensionality reduction

Regressing out the confounding effect of cell degradation (by modeling the mitochondrial percentage as a latent variable).

```{r dim_red_pca}
sc <- SCTransform(sc, vars.to.regress="percent.mito", verbose = FALSE) %>% RunPCA(verbose = FALSE)
ElbowPlot(sc, 50)
ggsave(paste0(plots_path, "elbow.pdf"))

sc$experiment <- ifelse(sc$orig.ident == "KA_B2" | sc$orig.ident == "KA_C11", "KA", "early_dev_d30")

if(params$harmonize) {
  sc <- RunHarmony(sc, assay.use = "SCT", group.by.vars = "experiment")
  p1 <- DimPlot(sc, reduction = "pca", group.by = "experiment")
  p2 <- DimPlot(sc, reduction = "harmony", group.by = "experiment")
  p <- p1 + p2 & ggmin::theme_powerpoint() & NoAxes()
  print(x = sc[['harmony']], dims = 1:5, nfeatures = 5, projected = FALSE)       
} else {
  p <- DimPlot(sc, reduction = "pca", group.by = "orig.ident") & ggmin::theme_powerpoint() & NoAxes()
  print(x = sc[['pca']], dims = 1:5, nfeatures = 5, projected = FALSE)
}
p
```


These are the genes driving the first 9 principal components

```{r dim_heatmap, fig.height=11}
if(params$harmonize) {
  DimHeatmap(sc, reduction = "harmony", dims = 1:9, cells = 500, balanced = TRUE)  } else {
  DimHeatmap(sc, dims = 1:9, cells = 500, balanced = TRUE)
}
```


# Clustering

Use UMAP for non-linear dimensionality reduction

Use Louvain algorithm for community detection.

```{r cluster_and_rdim, eval=TRUE, results=FALSE}
if(params$harmonize) {
  sc <- FindNeighbors(sc , reduction = "harmony", dims = 1:40, graph.name = "Harmony")
  sc <- FindClusters(sc, resolution = seq(1.5, 0.25, -0.25), graph.name = "Harmony") # reverse order, seurat_clusters takes last run's value
  sc <- FindClusters(sc, resolution = seq(0.20, 0.05, -0.05), graph.name = "Harmony") # reverse order, seurat_clusters takes last run's value
  sc <- RunUMAP(sc, reduction = "harmony", dims = 1:40)
  } else {
  sc <- FindNeighbors(sc, dims = 1:30)
  sc <- FindClusters(sc, resolution = seq(0.20, 0.05, -0.05)) # reverse order, seurat_clusters takes last run's value
  sc <- FindClusters(sc, resolution = seq(1.5, 0.25, -0.25)) # reverse order, seurat_clusters takes last run's value
  sc <- RunUMAP(sc, dims = 1:30)
}

```

Non-linear dimensionality reduction with UMAP

```{r plot_rdim}
DimPlot(sc, group.by = "orig.ident") + ggmin::theme_powerpoint()
ggsave(paste0(plots_path, "UMAP.color.by.orig.ident.pdf"), width = 8, height = 7)
```

Number of cells per cluster at different Louvain resolutions:

```{r}
if(params$harmonize) {
  clustree(sc, prefix = "Harmony_res.")
  } else {
  clustree(sc, prefix = "SCT_snn_res.")
}
```

# Cell-cycle stage classification

Score G1, G2/M and S signatures in each cell.

Classify phase based on that score.

```{r CC_regression, eval=TRUE}
if(params$cell.source == "human") {
  s.genes <- cc.genes.updated.2019$s.genes
  g2m.genes <- cc.genes.updated.2019$g2m.genes
} else {
  s.genes <- firstup(cc.genes.updated.2019$s.genes)
  g2m.genes <- firstup(cc.genes.updated.2019$g2m.genes)
}
sc <- CellCycleScoring(sc, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
grouping <- "Phase"; DimPlot(sc, reduction = "umap", group.by =  grouping) + ggmin::theme_powerpoint()

kable(table(sc$Phase, sc[[params$resDEG]][[1]]), caption="Cell cycle stage classification per cluster") %>%   kable_styling() %>%   scroll_box(width = "100%", height = "200px")
```

# Molecular Signatures scoring

Compute enrichment for molecular signatures using method from Tirosh et al, Science (2016)

```{r mol_sig_score, fig.height=11}
msig <- scan(paste0(cremwd, "/reference_data/gene_sets/lung.matur.differ.MSig.txt"), what="", sep="\n") # Read in the data
msig <- sapply(msig, function(k) gsub(",$", "", k))
msig <- strsplit(msig, ",[[:space:]]+")  # Separate elements by a comma and one or more whitepace
names(msig) <- sapply(msig, `[[`, 1)  # subset the first vector element and set it as the list element name
msig <- lapply(msig, `[`, -1) # Remove the first vector element from each list element
sc.markers.wilcox.res <- readRDS(paste0(cremwd, "project_workspace/19_09_10_darrell_habermann/calculations/analysis/habermann//rds/sc.markers.wilcox.celltype.Rds"))
sc.markers.wilcox.res[[2]]$cluster <- paste0("Hb_", sc.markers.wilcox.res[[2]]$cluster)  # habermann et al msig
msig2 <- split(sc.markers.wilcox.res[[2]]$gene, sc.markers.wilcox.res[[2]]$cluster)
msig <- c(msig, msig2)
names(msig) <- sub("\\+", "pos", names(msig))
names(msig) <- sub("-", "neg", names(msig))
names(msig) <- make.names(names(msig))
if(params$cell.source == "mouse") {
    msig <- lapply(msig, firstup)
}
sc <- AddModuleScore(sc, features = msig, name = names(msig), search = T)        
sc <- AddMetaData(sc, sc[[paste0(names(msig), 1:length(names(msig)))]], col.name 
= names(msig))
for(i in paste0(names(msig), 1:length(names(msig)))) {
  sc[[i]] <- NULL
}
```

# UMAP clusters

```{r facetted_louvain_clusters_UMAP, fig.height=6}
if(params$harmonize) {
  grouping <- "Harmony_res.0.1";  p1 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "Harmony_res.0.15"; p2 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "Harmony_res.0.2";  p3 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "Harmony_res.0.25"; p4 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "Harmony_res.0.5";  p5 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "Harmony_res.0.75"; p6 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  } else {
  grouping <- "SCT_snn_res.0.1";  p1 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "SCT_snn_res.0.15"; p2 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "SCT_snn_res.0.2";  p3 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "SCT_snn_res.0.25"; p4 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "SCT_snn_res.0.5";  p5 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
  grouping <- "SCT_snn_res.0.75"; p6 <- DimPlot(sc, group.by = grouping, label = T) + labs(title = grouping)
}
p <- p1 + p2 + p3 + p4 + p5 + p6
p <- p & ggmin::theme_powerpoint() & NoAxes() & NoLegend()
p
ggsave(paste0(plots_path, "UMAP.clust.pdf"), width = 10, height = 7)
```

# UMAP pre-selected features

```{r echo=FALSE}
# 4 columns, 9 width, height per line == 9 / 4 == 2.25
x <- interesting_feats <- c("nFeature_RNA", "nCount_RNA", "percent.mito", paste0("PC_", 1:10), names(msig), interesting_genes)
len <- (length(unique(x))/4)*2.25
```

```{r pre_selected_marker_genes_UMAP,  fig.height=len}
if(params$harmonize) {
  interesting_feats <- c("nFeature_RNA", "nCount_RNA", "percent.mito", paste0("harmony_", 1:10), names(msig), interesting_genes)
  } else {
  interesting_feats <- c("nFeature_RNA", "nCount_RNA", "percent.mito", paste0("PC_", 1:10), names(msig), interesting_genes)
}
p <- FeaturePlot(sc, features = interesting_feats, cols = c("#f0f0f0", brewer.pal(9,"OrRd")) )
p <- p & ggmin::theme_powerpoint() & NoAxes() & NoLegend()
p
ggsave(paste0(plots_path, "UMAP.preselected.feats.pdf"))
```


# Differential expression

Differential expression model: MAST, Finak et al.:
https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5       

For a recent comparison on the performance of different methods for single-cell differential expression, see:
https://www.nature.com/articles/nmeth.4612


```{r use_seurat, eval=T, fig.height=15}
fmas <- function(sc, res) {
    Idents(sc) <- res
    file_out <- paste0(calculations_path, "DEG.", res, ".xlsx")
    seurat.markers <- FindAllMarkers(sc, test.use = "MAST", only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, verbose = F)
    seurat.markers.summary <- as.data.frame(seurat.markers %>% group_by(cluster) 
%>% dplyr::filter(p_val_adj < 0.05) %>% top_n(20, abs(avg_logFC)))
    openxlsx::write.xlsx(seurat.markers, file_out, zoom = 100, asTable = TRUE, tableStyle = "none", rowNames = F)
    return(list(all = seurat.markers, top = seurat.markers.summary))
}
resDEG <- params$resDEG
sc.markers.mas.res <- fmas(sc, resDEG)
saveRDS(sc.markers.mas.res, paste0(rds_path, "sc.markers.mas.", resDEG, ".Rds")) 
```

Table of top 20 DEG per cluster

The clustering resolution used for differential expression is **`r resDEG`**     

```{r, eval=T}
sc.markers.mas.res[[2]] %>% dplyr::mutate(cluster = cell_spec(
    cluster, color = "white", bold = T,
    background = spec_color(as.numeric(factor(cluster)), end = 0.9, direction = -1)
  )) %>% kable(escape = F, caption=paste0("Top DEG for clusters at ", resDEG )) %>% kable_styling() %>% scroll_box(width = "100%", height = "600px")
```

# Heatmap of DEG

Heatmap of top 20 DEG per cluster

The clustering resolution used for differential expression is **`r resDEG`**     

```{r heatmap_DEG, eval = T, fig.height=15}
Idents(sc) <- resDEG
# switch to RNA normalized for heatmap visualization
DefaultAssay(sc) <- "RNA"
sc <- NormalizeData(sc)
all.genes <- rownames(sc)
sc <- ScaleData(sc, features = all.genes, vars.to.regress="percent.mito", verbose = F) # default: only scales var.genes, which misses some in heatmap
g <- DoHeatmap(sc, features = sc.markers.mas.res[[2]]$gene) + scale_fill_gradientn(colors = c("blue", "white", "red"))
print(g) # only need to use print if you aren't in a top-level statement (e.g. in a function, in an if/else block, etc.)
DefaultAssay(sc) <- "SCT"

```

```{r save_object}
saveRDS(sc, paste0(rds_path, "sc.Rds"))
```

# UMAP of DEG

The clustering resolution used for differential expression is **`r resDEG`**     

```{r echo=FALSE}
# 4 columns, 9 width, height per line == 9 / 4 == 2.25
x <- sc.markers.mas.res[[1]] %>% group_by(cluster) %>% dplyr::filter(p_val_adj < 
0.05) %>% top_n(4, avg_logFC) %>% pull(gene)
len2 <- (length(unique(x))/4)*2.25
```

```{r feat_DEG_UMAP,  fig.height=len2, error = TRUE}
# when allowing duplicates gives error
p <- FeaturePlot(sc, features = unique(x), cols = c("#f0f0f0", brewer.pal(9,"OrRd")) )
p <- p & ggmin::theme_powerpoint() & NoAxes() & NoLegend()
p
ggsave(paste0(plots_path, "UMAP.DEG.pdf"))
```

```{r export_spring, eval=T}
export.data <- GetAssayData(sc, slot = "counts")  # for SCT, counts are corrected for depth
export.genes <- rownames(sc)
write.table(as.matrix(export.data), gzfile(paste0(calculations_path, "spring.data.tsv.gz")), sep = "\t", quote = F, col.names = T, row.names = T )
write.table(export.genes, paste0(calculations_path, "spring.genes.tsv"), sep = "\t", quote = F, col.names = F, row.names = F )
if(params$harmonize) {
  meta.subset <- sc[[c("orig.ident", "Phase", grep("^Harmony", colnames(sc[[]]), 
value = T) )]]
  } else {
  meta.subset <- sc[[c("orig.ident", "Phase", grep("^SCT", colnames(sc[[]]), value = T) )]]
}
groupings <- t(meta.subset) # select metadata
meta.continuous <- sc[[names(msig)]]
custom.color.tracks <- t(meta.continuous)
write.table(custom.color.tracks, paste0(calculations_path, "spring.custom.color.tracks.csv"), sep = ",", quote = F, col.names = F, row.names = T )
write.table(groupings, paste0(calculations_path, "spring.groupings.csv"), sep = ",", quote = F, col.names = F, row.names = T )
```
